refactor(runtime): refactor filetype.lua

smjonas opened the pull request:
Some refactoring (fourth bullet point in #18604): move some fileteype detection functions from `filetype.lua` to `detect.lua`. These were introduced in #18219. Also sort the functions alphabetically like the existing ones.

smjonas committed:
refactor(runtime): refactor filetype.lua

Move some filetype detection functions to detect.lua,
sort patterns by detected filetype.

smjonas commented:
Yes, that makes sense. Btw, do you agree that all functions which check at least one line in the buffer can be moved to `detect.lua`? I find this more consistent than the rule "if more than line is checked, it should be in `detect.lua`".

clason commented:
Rebase? I think this can be tackled now, and should be independent of the `scripts.vim` refactor.

smjonas commented:
Yes, wasn't sure if #18895 would be merged first since that PR touches a lot of the same code as my PR. Should I just go ahead with refactoring?

clason commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -45,6 +45,17 @@ function M.asm(bufnr)
+function M.asp(bufnr)
+  if vim.g.filetype_asp then
+    return vim.g.filetype_asp
+  elseif table.concat(getlines(bufnr, 1, 3)):lower():find('perlscript') then
```
Would it make sense to add an optional argument (table) to the `getlines` function that allows passing `concat = true`?

clason replied:
Ah, OK. Then I defer to him -- never ping pong ;) (Changes shouldn't cycle; always optimize over the full span of the previous solutions so they improve on each.)

clason commented on runtime/lua/vim/filetype.lua:
```diff
@@ -1232,27 +1204,20 @@ local extension = {
-  end,
   -- Ignored extensions
   bak = function(path, bufnr)
     local root = vim.fn.fnamemodify(path, ':r')
```
would it make sense to extend `vim.fs.basename` to cover this? (@gpanders)

gpanders replied:
We could add `rootname` maybe? Though I don't think we need to add a `vim.fs` function for each fnamemodify variant. Personally I'm fine with sticking with fnamemodify here.

clason commented on runtime/lua/vim/filetype.lua:
```diff
@@ -1719,7 +1723,28 @@ local pattern = {
+    return require('vim.filetype.detect').dat(path, bufnr)
+  end,
+  ['[cC]hange[lL]og.*'] = starsetf(function(path, bufnr)
+    local line = M.getlines(bufnr, 1):lower()
```
same?

smjonas commented on runtime/lua/vim/filetype.lua:
```diff
@@ -685,6 +853,24 @@ local extension = {
+    return require('vim.filetype.detect').sh(path, bufnr, 'ksh')
+  end,
+  sh = function(path, bufnr)
+    return require('vim.filetype.detect').sh(path, bufnr, M.getlines(bufnr, 1))
```
Good idea, updated

smjonas commented on runtime/lua/vim/filetype.lua:
```diff
@@ -793,11 +979,7 @@ local extension = {
-    else
-      return 'typescript'
-    end
+    return M.getlines(bufnr, 1):find('<%?xml') and 'xml' or 'typescript'
```
Ok, I thought it might be simpler if we did not add a separate function for one-liners. But I guess that's not very consistent.

clason replied:
Yeah, that is a valid point. I was just going by the previous discussion (content check => detect.lua). We can come up with other rules; the main part is pick something workable and document it so it won't be a dice roll next time a pattern needs to be ported or implemented ;)

clason commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -327,6 +353,13 @@ function M.fs(bufnr)
 end
 
+function M.git(bufnr)
+  local line = etlines(bufnr, 1)
```
```suggestion
  local line = getlines(bufnr, 1)
```

gpanders commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -1055,6 +1086,18 @@ function M.web(bufnr)
 
+-- XFree86 config
+function M.xfree86(bufnr)
+  local line = getlines(bufnr, 1)
+  local on_detect
+  if matchregex(line, [[\<XConfigurator\>]]) then
+    on_detect = function(b)
+      vim.b[b].xf86conf_xfree86_version = 3
+    end
+  end
+  return 'xf86conf', on_detect
```
Might as well stuff the `matchregex` and `getlines` call into the `on_detect` function, as it doesn't affect the returned filetype at all.

```suggestion
  return 'xf86conf', function(b)
    local line = getlines(b, 1)
    if matchregex(line, [[\<XConfigurator\>]]) then
      vim.b[b].xf86conf_xfree86_version = 3
    end
  end
```

clason replied:
(could even do without the `local line`; just inline the `getlines(b ,1)`)

gpanders reviewed (approved):
Just one minor comment, otherwise LGTM

clason commented:
@gpanders Thoughts about keeping the "simple first line content checks" in `filetype.lua` vs. moving them to `detect.lua` (in light of your suggestion of separating  `bufnr`, `filename`, and `contents` in `match`)?

gpanders commented:
@clason and I have been discussing this on Matrix and we're both having difficulty coming up with a good heuristic for determining when a detection function should be inlined in `filetype.lua` versus `detect.lua`. My inclination would be to make the designation based on the "complexity" of the detection function: if the function is non-trivial or used more than once, it goes in `detect.lua`. Otherwise, just inline it. The problem with this is that "complexity" and "non-trivial" are both subjective and a bit arbitrary.

@smjonas If the goal is to organize these functions in a way that maximizes maintainability and minimizes cognitive overhead then I'd be very happy to solicit your input, given that you have spent as much time as anybody navigating around these functions.

clason commented:
That's a very good analysis. I would suggest
1. moving the functions you identified to `detect.lua` and
2. documenting your rule of thumb at the top of `detect.lua`, e.g.,
```lua
--  Filetype detection functions for use in `filetype.lua` that either
-- * are used more than once or
-- * check more than one line or use conditionals.
-- (Simple one-line checks, e.g., for a string in the first line are better inlined in filetype.lua.)
```
Then it's good to go 👍

smjonas commented:
I have added a comment to `detect.lua` and also refactored the function for the `printcap` check (which takes an argument for the `ptcap_type` now). I did not extract this function 
https://github.com/neovim/neovim/blob/29c43f7fde1ff354323673b828b353722d93fd97/runtime/lua/vim/filetype.lua#L2088
because it needs to call `filetype.lua`'s `match` function so I think it's ok like this.

clason commented:
Looks good to me! @gpanders one final look-see, then merge?

gpanders commented:
Looks good, thanks!

clason commented:
@smjonas again, thank you for the massive effort, patience, and attention to detail you put into this PR! We're getting closer and closer to the finish line for default `filetype.lua` :)

smjonas commented:
Awesome, glad I could help! :D One step closer to Lua filetype detection :rocket:
